Multi-Tenant Emails: Fixes for Retirement (Account Deletion)

OmarIthawi opened the pull request:
Please read and understand the [related decision](https://appsembler.atlassian.net/wiki/spaces/RT/pages/577569484/Decision+How+to+make+is+email+retired+function+multi-tenant) before reviewing this pull request.

### Related

 - This PR is part of the larger [Multi-Tenant Emails proposal](https://github.com/appsembler/tech-design-proposals/blob/master/proposals/0003-tahoe-multi-tenant.md).
 - Jira task: [MTE Retirement: As a learner registered in both Blue University and Green Academy I want to request the deletion of my account in Blue University and keep all my Green Academy untouched.](https://appsembler.atlassian.net/browse/RED-1017)


### TODO

 - [x] Waiting for [Decision: How to make is_email_retired function multi-tenant?](https://appsembler.atlassian.net/wiki/spaces/RT/pages/577569484/Decision+How+to+make+is+email+retired+function+multi-tenant)
- [x] Omar to add one more test case for the hack we’re proposing in Option 2 for this PR.
- [x] Omar to add follow up Jira tasks for the issues regarding cancel_user_retirement_request and Tubular.
  - See the [Pay Multi-Tenant Emails Tech Debt](https://appsembler.atlassian.net/browse/RED-1142) Epic
- [x] Review the pull request by @ahmedaljazzar 
- [x] Review the pull request by Maxi or John
- [x] Check whether the suffix log (i.e. turning `someone@example.com` -> `someone@example.com..org-short-name` for `UserRetirementStatus` has potential issues.
 - [x] Ahmed's request: `Can we test calling deactivate_user from within self.BLUE too?`

OmarIthawi commented on common/djangoapps/student/helpers.py:
```diff
@@ -69,6 +69,7 @@
     'generating',
     'downloadable',
 ]
+EMAIL_EXISTS_MSG_FMT = _("An account with the Email '{email}' already exists.")
```
Fixes a problem introduced by upstream and was only fixed until 24 days ago: 
 - https://github.com/edx/edx-platform/blame/cb0973317eb15ed7a5fb5e99fe249952f5b89ffc/common/djangoapps/student/helpers.py#L64

OmarIthawi commented on openedx/core/djangoapps/appsembler/sites/utils.py:
```diff
@@ -246,6 +246,30 @@ def get_current_organization(failure_return_none=False):
     return current_org
 
 
+@beeline.traced(name="get_current_organization_with_user")
+def get_current_organization_with_user(user, failure_return_none=False):
+    """
+    Get current organization from request using multiple strategies falling back to user organization.
+    """
+    request = get_current_request()
+
+    current_org = None
+    if request:
+        current_org = get_current_organization(failure_return_none=failure_return_none)
+
+    if not current_org:
+        # Falling back to the UserOrganizationMapping
+        try:
+            current_org = UserOrganizationMapping.objects.get(
+                user=user,
+            ).select_related('organization').organization
+        except UserOrganizationMapping.DoesNotExist:
+            if not failure_return_none:
+                raise
+
+    return current_org
+
+
```
```suggestion
```

OmarIthawi commented on openedx/core/djangoapps/appsembler/multi_tenant_emails/tests/test_utils.py:
```diff
@@ -16,7 +15,7 @@
 
 @contextlib.contextmanager
-def with_organization_context(site_color, configs=None):
+def with_organization_context(site_color, configs=None, fake_request=False):
```
```suggestion
def with_organization_context(site_color, configs=None):
```

OmarIthawi commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -426,11 +428,14 @@ def post(self, request):
                 UserSocialAuth.objects.filter(user_id=request.user.id).delete()
                 # Change LMS password & email
                 user_email = request.user.email
+
```
```suggestion
```

OmarIthawi commented on common/djangoapps/student/signals/receivers.py:
```diff
@@ -51,8 +53,27 @@ def on_user_updated(sender, instance, **kwargs):  # pylint: disable=unused-argum
+            organization = get_current_organization_with_user(user=instance, failure_return_none=True)
+            if is_email_retired(
+                    email=instance.email,
+                    check_within_organization=bool(organization),
```
```suggestion
                    # Avoid a chicken-and-egg problem in which that we don't have an active request to get the 
                    # organization from. 
                    #
                    # This specifically solves the issue with `student.signals..on_user_updated` for new users
                    # calls `is_email_retired` in non-request contexts.
                    #
                    # Chicken-and-egg: It's impossible to get the organization for a user that we didn't save yet
                    # to the database because we don't have `UserOrganizationMapping` yet.
                    # We cannot have `UserOrganizationMapping` until we save the user, therefore it's not possible to check.
                    check_within_organization=bool(organization),
```

OmarIthawi commented on common/djangoapps/student/signals/receivers.py:
```diff
@@ -51,8 +53,27 @@ def on_user_updated(sender, instance, **kwargs):  # pylint: disable=unused-argum
-            )
+        if not settings.FEATURES.get('APPSEMBLER_MULTI_TENANT_EMAILS', False):
+            # Note: Doing a loose check for retirement within organization in case
+            #       of non-request contexts such as management commands or Django shell.
```
```suggestion
            #       of non-request contexts such as management commands or Django shell which is not a 
            #       common use-case at all, except that it shows up in tests when we don't do full mocking.
            #
```

OmarIthawi commented on common/djangoapps/student/signals/receivers.py:
```diff
@@ -51,8 +53,27 @@ def on_user_updated(sender, instance, **kwargs):  # pylint: disable=unused-argum
+                    check_within_organization=bool(organization),
+                    organization=organization,
+            ):
+                raise AccountValidationError(
+                    EMAIL_EXISTS_MSG_FMT.format(username=instance.email),
+                    field="email"
+                )
```
This used to be a dead and broken code that never ran (as far as I know) on production. [It was fixed at upstream 24 days ago](https://github.com/edx/edx-platform/blame/cb0973317eb15ed7a5fb5e99fe249952f5b89ffc/common/djangoapps/student/helpers.py#L64).

However, I chose to go to the safer route and fix it and only check for organization-specific `is_email_retired` if have an organization in the request.

As described in the comments in non-request scenarios the fix is impossible since there's no way to know the organization before creating the `UserOrganizationMapping` which needs a user to exists. A chicken-and-egg scenario. 

So I went for a softer check that works only in contexts that has a `get_current_request` which we can infer the organization from.

OmarIthawi commented:
@ahmedaljazzar @johnbaldwin @melvinsoft this is ready for review. Please take a look and let me know what do you think.

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -117,6 +118,8 @@ class EnrollStatusChange(object):
 DEFAULT_TRANSITION_STATE = 'N/A'
 SCORE_RECALCULATION_DELAY_ON_ENROLLMENT_UPDATE = 30
 
+APPSEMBLER_BEFORE_RETIREMENT_EMAIL = 'APPSEMBLER_BEFORE_RETIREMENT_EMAIL'
```
Would you please explain this a bit with some inline code?

OmarIthawi replied:
Done

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
 
 
-def is_email_retired(email):
+def get_suffixed_email_for_retirement(email, organization):
```
Is there anywhere this will break email validation? IIRC, when a learner registers, the user's email is validated against a regex. We want to make sure that there isn't _somewhere_ in the system that making the domain invalid doesn't break something, like an async job because the email fails validation

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+
+    This fixes issues with the APPSEMBLER_MULTI_TENANT_EMAILS and is_email_retired.
+    """
+    suffixed_email = '{email}..{organization}'.format(
```
This isn't the whole email address, just the suffix, perhaps renaming the variable would help because when I read "suffxed_email" I think this the the whole email address, original + suffix

ahmedaljazzar replied:
can we replace the two-dots separator with something that doesn't belong to email patterns?

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
-def is_email_retired(email):
+def get_suffixed_email_for_retirement(email, organization):
+    """
+    Suffix the email with an unusable suffix with organization short_name.
```
Perhaps provide a "before" and "after" example in the docstring?

OmarIthawi replied:
done

johnbaldwin replied:
Thanks!

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+
+    This fixes issues with the APPSEMBLER_MULTI_TENANT_EMAILS and is_email_retired.
+    """
+    suffixed_email = '{email}..{organization}'.format(
```
Are these supposed to be dots or dashes?

OmarIthawi replied:
> 1. Are these supposed to be dots or dashes?

Good question. Dashes were dismissed because it might make a valid email suffix therefore allowing account hijacking. I've updated the decision page accordingly.

> 2. Is there anywhere this will break email validation? IIRC, when a learner registers, the user's email is validated against a regex. We want to make sure that there isn't _somewhere_ in the system that making the domain invalid doesn't break something, like an async job because the email fails validation

Yes, it could do that but it'll mostly break the retirement process in which should address in the [Multi-Tenant Emails Tech Debt Epic](https://appsembler.atlassian.net/browse/RED-1142) if any.

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+    profile.set_meta(meta)
+    profile.save()
+
+    user.email = get_suffixed_email_for_retirement(user.email, organization)
```
A couple of design points:

1. We've got two functions with very similar names. "suffix_user_email_for_retirements" and "get_suffixed_email_for_retirement". Just trying to think on how we can make this clearer as to what each does by the function names. Do you see where someone could think this is valid naming?

`user.email = suffix_user_email_for_retirement(...)` 

Why not just do the following?

`user.email += organization.shortname`

Unless we are unit testing the `get_suffixed_email_for_retirement` function in isolation. But I don't see any unit tests for this specific function...

So perhaps having this as a separate function is a bit of premature abstraction?

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+    """
+    suffixed_email = '{email}..{organization}'.format(
+        email=email,
+        organization=organization.short_name,
```
Is there any chance that `short_name` would have spaces or other invalid email characters? or does it not matter? Please see my comment above on email validation

OmarIthawi replied:
In theory, yes because we don't enforce a certain pattern at the model or database level.

But in general we add such checks in the trial signup workflow, I'll add a low priority task for follow up: https://appsembler.atlassian.net/browse/RED-1184

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
 
 
-def is_email_retired(email):
+def get_suffixed_email_for_retirement(email, organization):
```
This function is a variable constructor but has the context of a 'getter'. I would expect a 'getter' to retrieve data from either an internal state or from an external source. 

If we need to be explicit that we are adding a suffix to the email address, I think of something like `def add_retired_email_suffix(user, organization)`. If we do need to be explicit to our caller on our implementation detail is to add a suffix to an email, then why not just generate the suffix in this function and our callers do this:

```
user.email += retired_email_suffix(organization)
```
This way the function doesn't have to know anything about the user

However, I think that adding a suffix to the email in order to not get a collision is an implementation detail that our caller doesn't need to know about.

Let's say we've implemented this code and in acceptance testing or in production, we realized that instead of appending a string to the email, we actually want to hash the email, then encrypt it, or we realize we actually need a valid email address because a future version of Open edX enforces valid email addresses on saving. So now we need to rebuild the user's retired email address so it passes the email validator, or we need to modify the platform to loosen up the User.save() validation rules. Or something else we haven't foreseen.

So, I suggest we abstract this function with `generate_retired_email_address(user, organization)` which is both very specific in its context, but not its implementation details. And some future clever programmer doesn't make assumptions on how to be creative with this function. Just a thought.

I know that for this function, this isn't a big deal and in this specific instance, not critical. The function is only used in a couple of places for now.

OmarIthawi replied:
good idea.

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+    return suffixed_email
+
+
+def suffix_user_email_for_retirement(user):
```
Following my comment on `get_suffix_email_for_retirement`, I see this function does three things

A. create a new retired email address
B. Save the old email address in the user's profile metadata
C. Save the new state of the user

I looked at where this function is called and I see it is called in the deactivation API view:

https://github.com/appsembler/edx-platform/blob/omar/mte-delete-me/openedx/core/djangoapps/user_api/accounts/views.py#L417-L432

So how about renaming this function to `retire_user_email_address` or something like it because we're not just suffixing the email address. What we're doing at a higher context is retiring the email address and the rename abstracts internal details of how we retire the user's email address.

OmarIthawi replied:
Good idea, I've gone with `_retire_user_email_for_multi_tenancy`

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+
+    meta[APPSEMBLER_BEFORE_RETIREMENT_EMAIL] = user.email
+    profile.set_meta(meta)
+    profile.save()
```
To note, I see that the view that calls this wraps this call along with others in a transaction. Can we always expect this to be the case for anyone who calls this function? What if a new view calls this function and the new view does not wrap this call in a transaction?

OmarIthawi replied:
Good concern. I've moved this function to be a method within `DeactivateLogoutView` so it's no longer a concern, as `DeactivateLogoutView` always calls that function within a transaction.

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+    user.save()
+
+
+def is_email_retired(email, organization=None, check_within_organization=True):
```
love the function name!

OmarIthawi replied:
It's used:
![image](https://user-images.githubusercontent.com/645156/86820289-80af8980-c091-11ea-9fae-dc1547c0b00e.png)

also:

![image](https://user-images.githubusercontent.com/645156/86820349-97ee7700-c091-11ea-82aa-8d9f8c9b1307.png)

I'm leaning for explicit implementation in which I don't want to have `organization=None` mistakenly thought of as `check_within_organization=False`.

johnbaldwin replied:
Thanks!

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
+    if check_within_organization and settings.FEATURES.get('APPSEMBLER_MULTI_TENANT_EMAILS', False):
+        if not organization:
+            organization = get_current_organization()
+        suffixed_email = get_suffixed_email_for_retirement(email, organization)
```
Following from above comments, we could do `retired_email = generate_retired_email_address(email, organization)`

johnbaldwin commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +247,72 @@ def is_username_retired(username):
         settings.RETIRED_USER_SALTS,
         settings.RETIRED_EMAIL_FMT
     )
+    # At some point, we'll need to stop checking for the non-suffixed emails.
```
At what point do we stop checking?

OmarIthawi replied:
more context added

johnbaldwin commented on common/djangoapps/student/signals/receivers.py:
```diff
@@ -51,8 +53,39 @@ def on_user_updated(sender, instance, **kwargs):  # pylint: disable=unused-argum
+            #        For more information about the AccountValidationError below please checkout the
+            #        upstream conversations at:
+            #               - https://github.com/edx/edx-platform/pull/18136
+            #               - https://openedx.atlassian.net/browse/PLAT-2117
```
Thanks for adding the inline comments!

OmarIthawi commented:
> Hey @OmarIthawi Approving, but have some comments on framing the context and level of abstraction for the new functionality.
> 
> One question I'm not clear on. It looks like the code is retiring the user with the given username so that the human user can reuse their email address, but not the username. Just checking that this is correct

Thanks @johnbaldwin for the detailed review. I see several suggestions and I'm not sure which ones are a blocker and which ones are good-to-have?

I'll go over them shortly reply with what I know.

johnbaldwin commented:
@OmarIthawi Please add some strategic logging

OmarIthawi commented:
@johnbaldwin I've addressed all of your comments. I need to run the tests again. Thanks for the detailed review!

ahmedaljazzar commented on openedx/core/djangoapps/appsembler/multi_tenant_emails/tests/test_account_deletion.py:
```diff
@@ -0,0 +1,143 @@
+
+        with with_organization_context(site_color=self.BLUE):
+            blue_register_res = self.register_user(self.BLUE)
+            assert blue_register_res.status_code == status.HTTP_200_OK, blue_register_res.content
```
Can we test calling `deactivate_user` from within `self.BLUE` too?

ahmedaljazzar commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -469,6 +478,34 @@ def post(self, request):
+
+        organization = get_single_user_organization(user)
+
+        meta[self.APPSEMBLER_RETIREMENT_EMAIL_META_KEY] = user.email
```
According to the GDPR regulations, Article 5:

> The GDPR permits the storage of personal data for longer periods as long as that data is processed for archiving purposes (though it’s important to note that the law doesn’t specify minimum or maximum periods for data retention).

This gives us the advantage to keep such data as long as the user uses this email to access other accounts. But, let's say, the user deletes all of his accounts, this value should be scrapped as the user never existed.

OmarIthawi replied:
@ahmedaljazzar  

Thank you Mr. GDPR :smile:! I don't have the courage to go over the GDPR text!

It will be scrapped with the `profile.meta` field. You can verify that by checking a fully retired user in the admin panel. I'll share the link via Slack. So it should not be a concern.

This is kept in the case of a _cancelled_ account deletion. I'll add that as an inline comment.

ahmedaljazzar replied:
Perfect! Thanks @OmarIthawi

johnbaldwin replied:
@OmarIthawi Just thinking this through: Since we are not storing the email that has been marked as retired except in the User model, we don't have to be concerned about overwriting any existing value in `meta[self.APPSEMBLER_RETIREMENT_META_KEY]`, right?

And this is why I hold issue with functions that start with `get_` which are generators. Like: https://github.com/edx/user-util/blob/master/user_util/user_util.py#L48

`user_util.get_all_retired_emails` This implies it is getting them from _somewhere_ and not generating them. So we go looking for where it is pulling the data and, well, nope, not getting anything.

johnbaldwin replied:
@OmarIthawi one more thing. Maxi has a PR here that checks to make sure there are not any email conflicts: https://github.com/appsembler/amc/pull/358/files#diff-be625b23f5e6811ac1d2e1592a1c36edR42

Wondering if we want to create a model/table that either contains original emails or original email one way hashes to make sure that there is not a duplicate in AMC. Like for example that an AMC user requests an account to be deleted for GDPR. cc @melvinsoft

OmarIthawi replied:
> Perfect! Thanks @OmarIthawi

https://github.com/appsembler/edx-platform/blob/b81888323e8384096825e92d66522641813045f6/openedx/core/djangoapps/user_api/accounts/views.py#L990-L1000

https://github.com/edx/edx-platform/blob/b81888323e8384096825e92d66522641813045f6/openedx/core/djangoapps/user_api/accounts/views.py#L89

OmarIthawi commented on common/djangoapps/student/models.py:
```diff
@@ -244,17 +244,59 @@ def is_username_retired(username):
+
+    e.g. the email "omar@example.org" in blue-academy organization would become --> "omar@example.org..blue-academy"
+    """
+    retired_email = '{email}..{organization}'.format(
```
@ahmedaljazzar two dots makes an [invalid domain](https://webmasters.stackexchange.com/questions/119097/are-multiple-consecutive-dots-allowed-in-a-domain-name), so it should be fine for what you're asking for. Correct?

johnbaldwin replied:
Whichever technique we choose we will either have to stick with or we will need to implement follow up work to include the history of our "retired email encoding"

OmarIthawi commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -469,6 +478,34 @@ def post(self, request):
+
+        organization = get_single_user_organization(user)
+
+        meta[self.APPSEMBLER_RETIREMENT_EMAIL_META_KEY] = user.email
```
```suggestion
        # Store the original email in the case of Account Deletion cancellation is needed
        # This value will be removed when the pipeline is executed, so it won't 
        # degrade GDPR-compliance.
        meta[self.APPSEMBLER_RETIREMENT_EMAIL_META_KEY] = user.email
```

OmarIthawi commented on common/djangoapps/student/models.py:
```diff
@@ -273,7 +315,7 @@ def email_exists_or_retired(email, check_for_new_site=False):
     else:
         exists = User.objects.filter(email=email).exists()
-    return exists or is_email_retired(email)
+    return exists or is_email_retired(email, check_within_organization=not check_for_new_site)
```
Thanks for the comment. How does it look now @ahmedaljazzar?

```suggestion
    check_within_organization = not check_for_new_site # Allow learners to spin their new Tahoe trial
    return exists or is_email_retired(email, check_within_organization=check_within_organization)
```

OmarIthawi commented:
Thanks @ahmedaljazzar and @johnbaldwin for the detailed review. This PR is better with your notes :slightly_smiling_face:.

I think I've addressed all of your comments. If you'd like please a final look at it, before I merge it by Thursday COB.

ahmedaljazzar commented on openedx/core/djangoapps/appsembler/multi_tenant_emails/tests/test_account_deletion.py:
```diff
@@ -0,0 +1,146 @@
+            assert blue_register_res.status_code == status.HTTP_200_OK, blue_register_res.content
+            blue_deactivate_res = self.deactivate_user(self.RED)
+            assert blue_deactivate_res.status_code == status.HTTP_204_NO_CONTENT, blue_deactivate_res.content
+            assert UserRetirementStatus.objects.count() == 2
```
One interesting thing; we should've used `self.BLUE` instead of `self.RED` in [line#144](https://github.com/appsembler/edx-platform/pull/624/files#diff-608c1f8f28ae44b9c73e72e96e148e96R144), but we still got `UserRetirementStatus.objects.count() == 2` anyways, which is odd, it shows that a single user would still be able to retire from the same organization as many times as they wants, but that shouldn't be the case, right? I still have to check the pipeline if it would be able to handle this, but this should be prevented in the API imo. WDYT?

ahmedaljazzar replied:
OK. Taking my comment back, looks like the tests did there job by showing that we can't deactivate the user twice. Failing test here: https://travis-ci.org/github/appsembler/edx-platform/jobs/706231970

OmarIthawi replied:
lol! thanks Ahmed!

johnbaldwin commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -426,6 +433,8 @@ def post(self, request):
                 return verify_user_password_response
             with transaction.atomic():
+                if settings.FEATURES.get('APPSEMBLER_MULTI_TENANT_EMAILS', False):
+                    self._retire_user_email_for_multi_tenancy(request.user)
```
Thank you, @OmarIthawi !

johnbaldwin commented:
@OmarIthawi Thanks for making those changes! Almost there, just the failing test. Please address that and I'll approve

OmarIthawi commented on openedx/core/djangoapps/appsembler/multi_tenant_emails/tests/test_account_deletion.py:
```diff
@@ -0,0 +1,146 @@
+        with with_organization_context(site_color=self.BLUE):
+            blue_register_res = self.register_user(self.BLUE)
+            assert blue_register_res.status_code == status.HTTP_200_OK, blue_register_res.content
+            blue_deactivate_res = self.deactivate_user(self.RED)
```
```suggestion
            blue_deactivate_res = self.deactivate_user(self.BLUE)
```

OmarIthawi commented:
Thanks again @ahmedaljazzar and @johnbaldwin. Tests are fixed now.

cc: @melvinsoft if you'd like to take a look.
